eventmesh-connectors/eventmesh-connector-http/src/main/java/org/apache/eventmesh/connector/http/source/protocol/ProtocolFactory.java

```diff
@@ -0,0 +1,100 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.eventmesh.connector.http.source.protocol;
+
+import org.apache.eventmesh.connector.http.source.config.SourceConnectorConfig;
+import org.apache.eventmesh.connector.http.source.protocol.impl.CloudEventProtocol;
+import org.apache.eventmesh.connector.http.source.protocol.impl.CommonProtocol;
+import org.apache.eventmesh.connector.http.source.protocol.impl.GitHubProtocol;
+
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Protocol factory.
+ * This class is responsible for storing and creating instances of {@link Protocol} classes.
+ */
+public class ProtocolFactory {
+
+    // protocol name -> protocol class
+    private final ConcurrentHashMap<String, Class<?>> classes = new ConcurrentHashMap<>();
+
+    // protocol name -> protocol instance
+    private final ConcurrentHashMap<String, Protocol> instances = new ConcurrentHashMap<>();
+
+    private final SourceConnectorConfig sourceConnectorConfig;
+
+
+    public ProtocolFactory(SourceConnectorConfig sourceConnectorConfig) {
+        this.sourceConnectorConfig = sourceConnectorConfig;
+        // register all protocols
+        this.registerProtocol(CloudEventProtocol.PROTOCOL_NAME, CloudEventProtocol.class);
+        this.registerProtocol(GitHubProtocol.PROTOCOL_NAME, GitHubProtocol.class);
+        this.registerProtocol(CommonProtocol.PROTOCOL_NAME, CommonProtocol.class);
+    }
+
+
+    /**
+     * Register a protocol
+     *
+     * @param name  name of the protocol
+     * @param clazz class of the protocol
+     */
+    public void registerProtocol(String name, Class<?> clazz) {
+        name = name.toLowerCase(); // case insensitive
+        if (Protocol.class.isAssignableFrom(clazz)) {
+            classes.put(name, clazz);
+        } else {
+            throw new IllegalArgumentException("Class " + clazz.getName() + " does not implement Protocol interface");
+        }
+    }
+
+    /**
+     * Get an instance of a protocol, if it is not already created, create a new instance
+     *
+     * @param name name of the protocol
+     * @return instance of the protocol
+     */
+    public Protocol getInstance(String name) {
+        name = name.toLowerCase(); // case insensitive
+        if (instances.containsKey(name)) {
+            return instances.get(name);
+        }
+
+        if (classes.containsKey(name)) {
+            synchronized (this) {
+                // double check
+                if (instances.containsKey(name)) {
+                    return instances.get(name);
+                }
```

Comment:
Since the current thread might be waiting for other threads to release the lock before acquiring it, this operation is actually to prevent the situation where other threads have already created an instance but another instance is still created again. Of course, the concurrent competition mentioned above is just a precaution, or even an imagined scenario. Because `ProtocolFactory` is just a local variable in the current code, there won't be any concurrency safety issues. However, this "superfluous" action doesn't seem to have any worse impact.

Reply:
I will make variables and methods static and get rid of this useless concurrency design

Reply:
Resolved at https://github.com/apache/eventmesh/pull/4913/commits/fe74cc7b24676916cb40db98665d52b45918c3cd